compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java

```diff
@@ -4858,38 +4861,30 @@ private STNode parseNewKeyword() {
         }
     }
 
-    private STNode parseNewKeywordRhs(STNode newKeyword) {
-        STNode token = peek();
-        return parseNewKeywordRhs(token.kind, newKeyword);
-    }
-
     /**
      * <p>
      * Parse an implicit or explicit new expression.
      * </p>
      *
-     * @param kind       next token kind.
      * @param newKeyword parsed node for `new` keyword.
      * @return Parsed new-expression node.
      */
-    private STNode parseNewKeywordRhs(SyntaxKind kind, STNode newKeyword) {
-        switch (kind) {
-            case OPEN_PAREN_TOKEN:
-                return parseImplicitNewRhs(newKeyword);
-            case SEMICOLON_TOKEN:
-                break;
-            default:
-                if (isClassDescriptorStartToken(kind)) {
-                    return parseTypeDescriptorInNewExpr(newKeyword);
-                }
+    private STNode parseNewKeywordRhs(STNode newKeyword) {
+        STToken nextToken = peek();
+        if (nextToken.kind == SyntaxKind.OPEN_PAREN_TOKEN) {
+            return parseImplicitNewExpr(newKeyword);
         }
-        return STNodeFactory.createImplicitNewExpressionNode(newKeyword, STNodeFactory.createEmptyNode());
+        
+        if (isClassDescriptorStartToken(nextToken.kind)) {
+            return parseExplicitNewExpr(newKeyword);
+        }
+
+        return createImplicitNewExpr(newKeyword, STNodeFactory.createEmptyNode());
```

Comment:
Let's add recovery for NewKeyWordRhs. ErrorHandler already supports it. Recovery inside `parseClassDescriptor` will not hit because we call `parseExplicitNewExpr` only after seeing a class descriptor token. Let's move the recovery inside ``parseClassDescriptor` to `parseNewKeywordRhs`

Reply:
Adding recovery to new-keyword rhs is not possible. I tried. But, we can't predetermine all the new-keyword rhs.
e.g. tests like the below started to fail.
```
    RecordWithReadOnlyFields rec2 = {
        a: new AnotherReadOnlyClass(4567),
        c: new
    };
```

Reply:
Anyway, a recover point was added to parseClassDescriptor in case if we call it in an other place in furture.

Reply:
New keyword rhs tokens should be the tokens as implemented in errorhandler right?

Reply:
Resolving as per the discussion offline.